File: RoadRunnerJS/src/commonMain/kotlin/trajectory/Trajectory.kt
package trajectory

import geometry.Pose2d
import path.Path
import profile.MotionProfile
import kotlin.js.JsExport

/**
 * Trajectory backed by a [Path] and a [MotionProfile].
 *
 * @param path path
 * @param profile motion profile
 */
@JsExport
class Trajectory constructor(
    val path: Path,
    val profile: MotionProfile,
    val markers: Array<TrajectoryMarker> = emptyArray()
) {
    fun duration() = profile.duration()

    operator fun get(time: Double) = path[profile[time].x]

    fun velocity(time: Double): Pose2d {
        val motionState = profile[time]
        return path.deriv(motionState.x) * motionState.v
    }

    fun acceleration(time: Double): Pose2d {
        val motionState = profile[time]
        return path.secondDeriv(motionState.x) * motionState.v * motionState.v +
            path.deriv(motionState.x) * motionState.a
    }

    fun start() = path[0.0, 0.0]

    fun end() = path[path.length(), 1.0]
}

File: src/roadrunnerjs/trajectory/Trajectory.ts
import { Path } from "roadrunnerjs/path";
import { MotionProfile } from "roadrunnerjs/profile/MotionProfile";
import { TrajectoryMarker } from "./TrajectoryMarker";

/**
 * Trajectory backed by a [Path] and a [MotionProfile].
 *
 * @param path path
 * @param profile motion profile
 */
export class Trajectory {
  constructor(
    readonly path: Path,
    readonly profile: MotionProfile,
    readonly markers: TrajectoryMarker[] = []
  ) {}

  duration() {
    return this.profile.duration();
  }

  get(time: number) {
    return this.path.get(this.profile.get(time).x);
  }

  velocity(time: number) {
    const motionState = this.profile.get(time);

    return this.path.deriv(motionState.x).times(motionState.v);
  }

  acceleration(time: number) {
    const motionState = this.profile.get(time);

    return this.path
      .secondDeriv(motionState.x)
      .times(motionState.v)
      .times(motionState.v)
      .plus(this.path.deriv(motionState.x).times(motionState.a));
  }

  start() {
    return this.path.get(0, 0);
  }

  end() {
    return this.path.get(this.path.length(), 1);
  }
}
